codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -167,11 +167,29 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                 }
             }
 
-            // Pass headers to the flow-controller so it can maintain their sequence relative to DATA frames.
-            flowController().addFlowControlled(stream,
-                    new FlowControlledHeaders(stream, headers, streamDependency, weight, exclusive, padding,
-                            endOfStream, promise));
-            return promise;
+            // Trailing headers must go through flow control if there are other frames queued in flow control
+            // for this stream.
+            Http2RemoteFlowController flowController = flowController();
+            if (!endOfStream || !flowController.hasFlowControlled(stream)) {
```

Comment:
@ejona86 - The intent of `hasFlowControlled` is an optimization to bypass the flow controller if possible. We only need to queue headers in the flow controller if there are other frames (data frames) queued to preserve ordering. The optimization may let us "close" streams earlier with out having to wait for flow controller to get around to it. Do you see a case where the `1xx` message flow would break given the current logic?